### ISSUE TITLE
kubefed: Allow modifying startup flags for federation control plane

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#40398]
Right now admin cannot modify the startup flags used by `federation-apiserver` and `federation-controller-manager` while bringing them up using `kubefed init`.

We need to allow admins to do that.

cc @kubernetes/sig-federation-feature-requests @madhusudancs @irfanurrehman

**Comment by nikhiljindal** [comment]
The current workaround is to run `kubectl edit deployments federation-apiserver` after running `kubefed init`.

To implement this, we cannot add every apiserver and controller manager flag to kubefed.
We need a generic way where the flags are opaque to kubefed. kubefed should just pass the flags as is to the apiserver or controller manager.

**Comment by irfanurrehman** [comment]
I would take this up if its ok. (would be able to work on it monday onwards)
@nikhiljindal @madhusudancs

**Comment by madhusudancs** [comment]
@irfanurrehman go for it! Assigned the issue to you.

**Comment by madhusudancs** [comment]
@irfanurrehman also, please discuss before implementing the solution.

**Comment by marun** [comment]
@madhusudancs Would it make sense to take this opportunity to move away from flags and start using ComponentConfig instead?  This approach is being proposed for components like the kubelet, and as per some of the discussion on #29459 there are some advantages that may be worth considering. Some advantages of ComponentConfig include:

 - typing and versioning (vs flags which are neither)
 - enables rolling out configuration changes to a running control plane

**Comment by madhusudancs** [comment]
@marun ComponentConfig is valuable and is the way to go in medium-to-long term. There are other challenges that come with it, particularly around dealing with dynamic updates. Nobody has committed to getting it done anytime soon, but we need something working soon. And passthrough is not a bad short-to-medium term solution until ComponentConfig materializes.

**Comment by nikhiljindal** [comment]
Even with `ComponentConfig`, we need to figure out how can a user edit `federation-apiserver` flags while bringing it up with `kubefed init`. User needs to pass the flags to kubefed to be able to modify the `ComponentConfig` that `kubefed` creates.

So we can move to ComponentConfig later while retaining the mechanism we come up here of users passing flags to kubefed?

**Comment by irfanurrehman** [comment]
@nikhiljindal @madhusudancs 
I am planning to implement this using flags --apiserver-args-override and --ctrlmgr-args-override
and values to each flag being an array of csv entries, each entry delimited by a newline (as is the case with csv files).
for example"
```
kubefed init <other flags> --apiserver-args-override "--bind-address,10.1.2.3
                                                    --insecure-bind-address,192.168.1.2"
```
If needed we can then further enhance this to take the pairs from a csv file or provide both options.

I also looked up the usage of config (which is used in many commands, for example dns provide config).
I think it should be equally simple for the user to specify the ini config in place of csv.
for example"  
```
kubefed init <other flags> --apiserver-args-override "apiserverflags
                                                 bind-address=10.1.2.3
                                                 insecure-bind-address=192.168.1.2"
```
This also can be extended to take an actual config file from user later.
A sample usage is here https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/providers/coredns/coredns.go#L59

I dont have any inclination to either, the csv approach looks like lesser code though. 
Please let me know your thoughts. Also if you have a better alternative.

**Comment by marun** [comment]
I don't think csv is a good idea for a file format.  yaml and json are what are used today in kube, and I don't think there's a good reason to deviate from that for ``kubefed``.

For the command line argument, I think csv makes even less sense.  Consistency is important, and I'm not aware of any kube command that uses csv in this way.  How about allowing flags to be provided in the same form as in they appear in the deployment template?

```bash
kubefed init ... --apiserver-arg-overrides="--bind-address=0.0.0.0,--secure-port=443"

# I think using space as a separator reads better, but if argument values 
# include spaces the nested quoting could get messy.
kubefed init ... --apiserver-arg-overrides="--bind-address=0.0.0.0 --secure-port=443"
```

**Comment by madhusudancs** [comment]
I agree. I am Ok not having a file format at all for now. Hopefully `ComponentConfig` will fix that. But command line arguments should not be CSV and even more so should not be line separated.

@marun's suggestion is a better alternative.

**Comment by madhusudancs** [comment]
I am in favor of comma separation instead of space separation. Space separation is hard to read if you are supplying a long list of overrides. "--bind-address=0.0.0.0 --tls-ca-file=/path/to/file --tls-cert-file=/path/to/file --tls-private-key-file=/path/to/file --requestheader-client-ca-file=/path/to/file --secure-port=443". Now if I want to distinguish between kubefed's flags and the overrides, I have to search for the quotes. On the other hand commas are not used for flag separation in kubefed and are hence easy to distinguish.

**Comment by irfanurrehman** [comment]
@marun @madhusudancs , thanks for your suggestions.
My only inclination for the way I did suggest was for the reason of trying to keep it compatible with a future file format. (I did read some discussions which indicated that taking these from a config might be more needed in future). If we keep that aside @marun's suggestion makes perfect sense.
I will implement this on the same lines.

**Comment by madhusudancs** [comment]
@irfanurrehman I think it is hard to keep the overrides, the way it is suggested above, compatible with the future file formats. We don't use CSV anywhere and unlikely we are going to adopt it. So it is not going to be compatible that way.

Also, I would like @nikhiljindal to weigh in before implementing this.

**Comment by madhusudancs** [comment]
@irfanurrehman I discussed this with @nikhiljindal in-person and he is Ok with your proposal. But let's have a comma separated list as suggested by @marun in https://github.com/kubernetes/kubernetes/issues/40398#issuecomment-276522846.

**Comment by nikhiljindal** [comment]
Yes csv is better. thanks for checking!

**Comment by madhusudancs** [comment]
Just to clarify, by csv @nikhiljindal meant the comma separation, not the line separation in the initial proposal.

**Comment by nikhiljindal** [comment]
yes :)

**Comment by irfanurrehman** [comment]
yes I got that.. :)

**Comment by madhusudancs** [comment]
Fixed by @irfanurrehman in PR #40917.
